nilearn/connectome/tests/test_connectivity_matrices.py

```diff
@@ -306,327 +388,463 @@ def test_geometric_mean_properties():
+    ):
+        conn_measure.fit(1.0)
 
-    # Raising error for input subjects with different number of features
-    pytest.raises(
-        ValueError, conn_measure.fit, [np.ones((100, 40)), np.ones((100, 41))]
-    )
+    # input subjects not 2D numpy.ndarrays
+    with pytest.raises(
+        ValueError, match="Each subject must be 2D numpy.ndarray."
+    ):
+        conn_measure.fit([np.ones((100, 40)), np.ones((10,))])
 
-    # Raising an error for fit_transform with a single subject and
-    # kind=tangent
+    # input subjects with different number of features
+    with pytest.raises(
+        ValueError, match="All subjects must have the same number of features."
+    ):
+        conn_measure.fit([np.ones((100, 40)), np.ones((100, 41))])
+
+    # fit_transform with a single subject and kind=tangent
     conn_measure = ConnectivityMeasure(kind="tangent")
-    pytest.raises(
+
+    with pytest.raises(
         ValueError,
-        conn_measure.fit_transform,
-        [np.ones((100, 40))],
-    )
+        match="Tangent space parametrization .* only be .* group of subjects",
+    ):
+        conn_measure.fit_transform([np.ones((100, 40))])
 
 
-def test_connectivity_measure_outputs():
-    n_subjects = 10
-    n_features = 49
+@pytest.mark.parametrize(
+    "cov_estimator", [EmpiricalCovariance(), LedoitWolf()]
+)
+@pytest.mark.parametrize("kind", CONNECTIVITY_KINDS)
+def test_connectivity_measure_generic(
+    kind, cov_estimator, signals_and_covariances
+):
+    signals, covs = signals_and_covariances
 
-    # Generate signals and compute covariances
-    emp_covs = []
-    ledoit_covs = []
-    signals = []
-    ledoit_estimator = LedoitWolf()
-    for k in range(n_subjects):
-        n_samples = 200 + k
-        signal, _, _ = generate_signals(
-            n_features=n_features,
-            n_confounds=5,
-            length=n_samples,
-            same_variance=False,
+    # Check outputs properties
+    input_covs = copy.copy(covs)
+    conn_measure = ConnectivityMeasure(kind=kind, cov_estimator=cov_estimator)
+    connectivities = conn_measure.fit_transform(signals)
+
+    # Generic
+    assert isinstance(connectivities, np.ndarray)
+    assert len(connectivities) == len(covs)
+
+    for k, _ in enumerate(connectivities):
+        assert_array_equal(input_covs[k], covs[k])
+
+        assert is_spd(covs[k], decimal=7)
+
+
+def _assert_connectivity_tangent(connectivities, conn_measure, covs):
+    for true_covariance_matrix, estimated_covariance_matrix in zip(
+        covs, connectivities
+    ):
+        """Check output value properties for tangent connectivity measure \
+        that they have the expected relationship \
+        to the input covariance matrices.
+
+        - the geometric mean of the eigenvalues
+          of the mean covariance matrix is positive-definite
+        - the whitening matrix (used to transform the data \
+          also produces a positive-definite matrix
+        """
```

Comment:
Adding some plain english explanation here but I would not mind having someone cross check that I want I am adding is correct: my linear algebra is not what it used to be.

Reply:
Note that some of the steps in this functions are not explained in the doc string: mostly because I am not even entirely sure what the test actually does.